### PR TITLE
Update shape inference documentation

### PR DESCRIPTION
**Pull request by gramalingam**
Signed-off-by: Ganesan Ramalingam <grama@microsoft.com>

**Description**
Update the shape-inference documentation.

**Motivation and Context**
Helps developers who write shape-inference code for operators avoid some common mistakes.

### CODE REVIEW

File: docs/ShapeInference.md [hunk 1]
```diff
@@ -52,22 +87,30 @@ To see numerous examples, search for occurrences of
...
+
+There are several utility functions in [shape_inference.h](/onnx/defs/shape_inference.h)
+to handle various common situations.
+
+* Use `checkInputRank` for inputs that must have a fixed rank. (See the
+inference for `RoiAlign` as an example.)
+
+* `unifyInputDim` and `unifyDim` and `updateOutputShape` can be used
+when multiple input dims are expected to be the same, and when input
+dimensions are propagated to specific output dimensions.(See the inference
```
Comment (jcwchen) [inline docs/ShapeInference.md:112]:
```suggestion
dimensions are propagated to specific output dimensions. (See the inference
```
nit

**Review by jcwchen — COMMENTED** [review]
I like this document enhancement -- especially the documentation for shape inference utilities.

File: docs/ShapeInference.md (hunk 1 above)
Comment (gramalingam) [reply docs/ShapeInference.md]: Added, thanks!

**Review by jcwchen — APPROVED** [review]

File: docs/ShapeInference.md [hunk 2]
```diff
@@ -1,5 +1,20 @@
 <!--- SPDX-License-Identifier: Apache-2.0 -->
 
+# Background
```
Comment (xadupre) [inline docs/ShapeInference.md:3]: Only one main title or several?

File: docs/ShapeInference.md [hunk 3]
```diff
@@ -52,22 +87,51 @@ To see numerous examples, search for occurrences of
...
+tensor whose rank is unknown and handled appropriately. Usually, the
+shape-inference logic is guarded by a call to `hasInputShape` or
+`hasNInputShapes`.
+
+* Before accessing the `dim_value` or `dim_param` of any dimension, the
+code must check if these fields have a value. In particular, the code must
+handle the possibility that the dimension may not have a statically
+known value.
+
+There are several utility functions in [shape_inference.h](/onnx/defs/shape_inference.h)
```
Comment (xadupre) [inline docs/ShapeInference.md:104]: Should we make sure users are using the utility functions to throw exceptions if one shape has an unexpecgted value?

File: docs/ShapeInference.md (hunk 2 above)
Comment (gramalingam) [reply docs/ShapeInference.md]: Made into a sub-section and moved it, thanks!

File: docs/ShapeInference.md (hunk 3 above)
Comment (gramalingam) [reply docs/ShapeInference.md:104]: Yes, but, unfortunately, we are dependent on reviewers to check for this. Any other suggestions?

**Pull request merged**